src/types.ts

```diff
@@ -0,0 +1,356 @@
+import type { ReadableStream, ReadableWritablePair } from 'stream/web';
+import type { ContextTimed, ContextTimedInput } from '@matrixai/contexts';
+import type { Id } from '@matrixai/id';
+import type { Caller } from './callers';
+import type { RawCaller } from './callers';
+import type { DuplexCaller } from './callers';
+import type { ServerCaller } from './callers';
+import type { ClientCaller } from './callers';
+import type { UnaryCaller } from './callers';
+import type Handler from './handlers/Handler';
+
+/**
+ * This is the JSON RPC request object. this is the generic message type used for the RPC.
+ */
+type JSONRPCRequestMessage<T extends JSONValue = JSONValue> = {
+  /**
+   * A String specifying the version of the JSON-RPC protocol. MUST be exactly "2.0"
+   */
+  jsonrpc: '2.0';
+  /**
+   * A String containing the name of the method to be invoked. Method names that begin with the word rpc followed by a
+   *  period character (U+002E or ASCII 46) are reserved for rpc-internal methods and extensions and MUST NOT be used
+   *  for anything else.
+   */
+  method: string;
+  /**
+   * A Structured value that holds the parameter values to be used during the invocation of the method.
+   *  This member MAY be omitted.
+   */
+  params?: T;
+  /**
+   * An identifier established by the Client that MUST contain a String, Number, or NULL value if included.
+   *  If it is not included it is assumed to be a notification. The value SHOULD normally not be Null [1] and Numbers
+   *  SHOULD NOT contain fractional parts [2]
+   */
+  id: string | number | null;
```

Comment:
Lets make id default to `null`

Allow the user to pass in a generator function that returns `string | number | null` in case they want to use a specific id for debugging purposes. We're going to be doing this in PK

Reply:
wdym?

Reply:
Set the `id` to default to `null`.

Then in `RPCServer` and `RPCClient`, accept a `idGen` function as a parameter that when called gives you a `string | number | null` to put into the ID of every message sent out.

```ts
idGen: () => PromiseLike<string | number | null>
```

Then do `await idGen()`.

Later can be extended to take various input parameters.